domain/port/state/state.go

```diff
@@ -0,0 +1,282 @@
+		if err != nil {
+			return domain.CoerceError(err)
+		}
+		for _, protocol := range protocols {
+			protocolMap[protocol.Name] = protocol.ID
+		}
+
+		// Get the current opened ports for the unit's endpoints.
+		// This will be used to calculate the changes required.
+		var currentOpenedEndpointPorts []endpointPortRange
+		err = tx.Query(ctx, getOpenedPortsOnEndpoints, unitUUID, endpointsUnderAction).GetAll(&currentOpenedEndpointPorts)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			currentOpenedEndpointPorts = []endpointPortRange{}
+		} else if err != nil {
+			return domain.CoerceError(err)
+		}
+
+		endpointMap := map[string][]portRange{}
+		for _, r := range currentOpenedEndpointPorts {
+			endpointMap[r.Endpoint] = append(endpointMap[r.Endpoint], portRange{
+				Protocol: r.Protocol,
+				FromPort: r.FromPort,
+				ToPort:   r.ToPort,
+			})
+		}
+
+		unitPortRanges := []unitPortRange{}
+		changedEndpoints := endpointUUIDs{}
+
+		// Calculate the end-state of the ports for each endpoint, for endpoints
+		// that will be affected by the action. If no changes are required, the
+		// endpoint will not be included.
+		for endpoint, endpointUUID := range endpointsUnderActionToUUID {
+			currentOpenedPorts := endpointMap[endpoint]
+
+			currentNetworkPorts := network.NewPortRanges(transform.Slice(currentOpenedPorts, decodePortRange)...)
+			reconciledOpenPorts := reconcilePorts(currentNetworkPorts, openPorts[endpoint], closePorts[endpoint])
+
+			if !currentNetworkPorts.EqualTo(reconciledOpenPorts) {
+				changedEndpoints = append(changedEndpoints, endpointUUID.UUID)
+				unitPortRanges = append(unitPortRanges, transform.Slice(reconciledOpenPorts, func(p network.PortRange) unitPortRange {
+					return unitPortRange{
+						ProtocolID:       protocolMap[p.Protocol],
+						FromPort:         p.FromPort,
+						ToPort:           p.ToPort,
+						UnitEndpointUUID: endpointUUID.UUID,
+					}
+				})...)
+			}
+		}
+
+		// Clear out all existing open ports for endpoint undergoing changes.
+		if len(changedEndpoints) > 0 {
+			err = tx.Query(ctx, clearPortRange, changedEndpoints).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+		}
+
+		// Insert the freshly calculated port ranges
+		if len(unitPortRanges) > 0 {
+			err = tx.Query(ctx, insertPortRange, unitPortRanges).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+		}
+
+		return nil
+	})
+}
+
+// reconcilePorts returns the reconciled list of port ranges after applying the
+// operations to open and close various additional port ranges.
+func reconcilePorts(currentOpened network.PortRanges, openPorts, closePorts []network.PortRange) network.PortRanges {
+	portRanges := currentOpened
+	for _, openRange := range openPorts {
+		portRanges = portRanges.Add(openRange)
+	}
+	for _, closeRange := range closePorts {
+		portRanges = portRanges.Remove(closeRange)
+	}
+	return portRanges
+}
+
+func decodePortRange(p portRange) network.PortRange {
```

Comment:
Might as well make these methods of the portRange type and put them in the types.go file, thats a pattern we use elsewhere.

Reply:
You can still use them with the slices, will just have to do `(portRange).decode` rather than `decodePortRange`

Reply:
Oh I didn't know you could do that!